File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.java
import java.util.Queue;
import java.util.LinkedList;

class Solution {
    public int solution(int[][] maps) {
        int answer = 0;
        
        answer = bfs(maps);
        return answer;
    }
    
    public int bfs(int[][] maps){
        Queue<Cord> queue = new LinkedList<>();
        int n = maps.length;
        int m = maps[0].length;
        maps[n-1][m-1] = -1;
        boolean[][] visited = new boolean[n][m];
        queue.add(new Cord(0,0));
        visited[0][0] = true;
        int[] dx = {1,0,-1,0};
        int[] dy = {0,1,0,-1};
        
        while(!queue.isEmpty()){
            Cord now = queue.poll();
            if(now.x == n-1 && now.y == m-1) break;
            for(int i=0;i<4;i++){
                int nx = now.x + dx[i];
                int ny = now.y + dy[i];
                
                if(nx>=0 && nx<n && ny>=0 && ny<m){
                    if(!visited[nx][ny] && maps[nx][ny] != 0){
                        queue.add(new Cord(nx,ny));
                        visited[nx][ny] = true;
                        maps[nx][ny] = maps[now.x][now.y]+1;
                    }
                }
            }
        }
        return maps[n-1][m-1];
    }
    
    class Cord{
        int x;
        int y;
        
        public Cord(int x, int y){
            this.x=x;
            this.y=y;
        }
    }
}

File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.py
from collections import deque

dx = [0,1,0,-1]
dy = [1,0,-1,0]

def solution(maps):
    answer = 0
    queue = deque()
    n,m=len(maps), len(maps[0])
    
    visited = [[False for _ in range(m)] for _ in range(n)]
    visited[0][0] = True
    queue.append([0,0])
    
    while queue:
        x,y = queue.popleft()
        
        for i in range(4):
            nx, ny= x+dx[i], y+dy[i]
            
            if 0<=nx<n and 0<=ny<m and not visited[nx][ny] and maps[nx][ny] != 0:
                visited[nx][ny] = True
                maps[nx][ny] = maps[x][y] +1
                queue.append([nx,ny])
            if nx == n-1 and ny == m-1:
                return maps[nx][ny]
    
    return -1
    